src/Network/Broadcast/OutboundQueue.hs

```diff
@@ -1032,32 +1044,33 @@ flush OutQ{..} = liftIO $ do
   queue again.
 -------------------------------------------------------------------------------}
 
--- | Update the set of peers
---
--- NOTE: Any messages to removed peers will be deleted from the queue.
+newtype Bucket nid = Bucket { bucketMVar :: MVar (Peers nid) }
+
+-- | Internal method: read all buckets of peers
+getAllPeers :: MonadIO m => OutboundQ msg nid buck -> m (Peers nid)
+getAllPeers OutQ{..} = liftIO $
+    mconcat <$> mapM (readMVar . bucketMVar) (Map.elems qBuckets)
+
+-- | Update a bucket of peers
 --
--- TODO check for overlap in the Peers terms. This should never happen but
--- could, due to misconfiguration.
-updateKnownPeers :: MonadIO m
-                 => OutboundQ msg nid -> (Peers nid -> Peers nid) -> m ()
-updateKnownPeers OutQ{..} f = do
-    removed <- applyMVar qPeers $ \peers ->
-      let peers'  = f peers
-          removed = peersToSet peers Set.\\ peersToSet peers'
-      in (peers', removed)
+-- NOTE: Any messages to peers that no longer exist in _any_ bucket will be
+-- removed from the queue.
+updatePeersBucket :: MonadIO m
+                  => OutboundQ msg nid buck
+                  -> buck
+                  -> (Peers nid -> Peers nid)
+                  -> m ()
+updatePeersBucket outQ@OutQ{..} buck f = do
+    before <- getAllPeers outQ
+    let Bucket bucket = qBuckets Map.! buck
```

Comment:
I started with qBuckets being an `MVar` but I removed it: it's much easier to understand the behaviour of the queue when the set of buckets is entirely static (and ordered).